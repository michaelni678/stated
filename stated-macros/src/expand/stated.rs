use proc_macro2::TokenStream as TokenStream2;
use quote::format_ident;
use syn::{ItemImpl, ItemStruct, Meta, Result, Token, punctuated::Punctuated};

use crate::{
    extensions::ty::{TypeExt, TypePathExt},
    utilities::squote::squote,
};

pub fn expand_item_struct(
    metas: Punctuated<Meta, Token![,]>,
    item_struct: ItemStruct,
) -> Result<TokenStream2> {
    let macro_name = format_ident!("__{}", item_struct.ident);

    Ok(squote! {
        // Re-emit the struct with the internal macro.
        #[::stated::stated_internal(#metas)]
        #item_struct

        #[doc(hidden)]
        macro_rules! #macro_name {
            ($($tt:tt)*) => {
                // Re-emit the input, but with the metas attached.
                #[::stated::stated_internal(#metas)]
                $($tt)*
            }
        }

        // Make the macro public with the same name as the struct.
        pub(crate) use #macro_name as #{item_struct.ident};
    })
}

pub fn expand_item_impl(item_impl: ItemImpl) -> Result<TokenStream2> {
    // Expect a macro at the impl type path with the same name.
    let mut macro_path = item_impl.self_ty.require_path()?.clone();

    // Strip the generic arguments from the macro path.
    macro_path.strip_generics();

    Ok(squote! {
        #macro_path!(#item_impl);
    })
}
