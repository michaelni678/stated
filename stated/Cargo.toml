[package]
name = "stated"
version = "0.1.0"
edition = "2024"
authors = ["Michael Ni <michael.ni678@gmail.com>"]
repository = "https://github.com/michaelni678/stated"
homepage = "https://github.com/michaelni678/stated"
readme = "README.md"
license = "Apache-2.0"
keywords = ["typestate"]
description = "Typestate pattern, made simple"
documentation = "https://docs.rs/stated"

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]

[features]
guide = []

[dependencies]
stated-macros = { path = "../stated-macros", version = "0.1.0" }
