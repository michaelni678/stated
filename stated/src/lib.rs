#![cfg_attr(docsrs, feature(doc_auto_cfg))]

//! Stated simplifies working with the typestate pattern.

extern crate self as stated;

pub use stated_macros::{stated, stated_internal};

#[cfg(feature = "guide")]
pub mod guide;

/// Indicates a disabled state.
#[derive(Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct N;

/// Indicates an enabled state.
#[derive(Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Y;

/// Placeholder for a stateless type.
pub struct __;
