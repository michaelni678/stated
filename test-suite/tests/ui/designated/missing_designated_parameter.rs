use stated::stated;

#[stated(states(A))]
pub struct Test<S>;

#[stated]
impl<S> Test<S> {}

fn main() {}
