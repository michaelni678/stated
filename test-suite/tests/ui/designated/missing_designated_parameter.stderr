error: no parameter is designated
 --> tests/ui/designated/missing_designated_parameter.rs:4:17
  |
4 | pub struct Test<S>;
  |                 ^

error: no parameter is designated
 --> tests/ui/designated/missing_designated_parameter.rs:7:6
  |
7 | impl<S> Test<S> {}
  |      ^
